### ISSUE TITLE
FCodePrinter and Integer arguments

### ISSUE THREAD

**Issue by alpxp** [sympy/sympy#20435]
How can I overcome the error which this code generates:

```
import sympy as sp
from sympy.utilities.autowrap import autowrap
from sympy.printing.fcode import FCodePrinter
from sympy.utilities.codegen import FCodeGen
x = sp.Symbol('x')
n = sp.Symbol('n', integer= True)
exp = sp.besselj(n+1,x**2)
printer = FCodePrinter(settings={'user_functions':{'besselj':'BESJN'}})
besjn = autowrap(exp, args=[n,x], backend="f2py", code_gen=FCodeGen(printer=printer),tempdir='autowraptmp')
```
The error is
```wrapped_code_0.f90:14:17:
   14 | autofunc = BESJN(n + 1.0d0, x**2)
      |                 1
Error: 'n' argument of 'besjn' intrinsic at (1) must be INTEGER
```
which due to FCodePrinter replacing an integer in the function argument with the float.
Removing +1 from `exp = sp.besselj(n+1,x**2)` is working fine.

**Comment by bjodah** [comment]
Thank you for reporting this.

I'm guessing the 1.0d0 comes from:
https://github.com/sympy/sympy/blob/c0e85160406f9bf2bcaa2992138587668a1cd0bc/sympy/printing/fortran.py#L368

We need to fix this in SymPy. You can try subclassing FCodePrinter and provide a `_print_Integer` method (but it's guesswork on my part, I haven't tried it).

**Comment by alpxp** [comment]
@bjodah Do you mean something like this?
```
from sympy.printing.codeprinter import CodePrinter
class CustomPrinter(FCodePrinter):
    def _print_Integer(self, expr):
        return CodePrinter._print_Integer(self, expr)
printer = CustomPrinter(settings={'user_functions':{'besselj':'BESJN'}})
```
Did not help.
But I'm reading:
https://github.com/sympy/sympy/blob/c0e85160406f9bf2bcaa2992138587668a1cd0bc/sympy/printing/fortran.py#L302
Is this the problem?

**Comment by bjodah** [comment]
Nice catch! Yes it has got to be. I'm thinking we should teach the printer about argument types of Fortran's builtin functions, and then probably provide a printer options which users can pass to inform the printer of argument types of external functions.

**Comment by alpxp** [comment]
Ok. I see. 

> provide a printer options which users can pass to inform the printer of argument types of external functions.

I think this is a good idea.
I was puzzled, btw, that `besselj` can not be autowraped by default. But it's a separate issue. I thought I'd workaround it via `user_functions`, but that failed as well (as per above issue).

**Comment by bjodah** [comment]
Yes, in the meantime you can override `_print_Function`, a pull request is most welcome if you have the time and feel like it.

**Comment by alpxp** [comment]
What do you have in mind for the pull request:
- Implementing `besselj`?
- implementing a new interface to specify argument types for `user_functions` ?
or
- changing the current behaviour so that the argument types are preserved?
I can look at the last point, but I'm currently not certain how to implement it. Any hints?

**Comment by bjodah** [comment]
`besselj` is part of Fortran since 2007? I think it would be great if the fortran printers, out of the box, support all SymPy constructs which have a (standard) Fortran counterpart.
And next logical step is of course to make use of 3rd party functions as smooth as possible. Exactly what this means is perhaps best answered by someone who uses SymPy and fortran, so I don't dare say for sure.

**Comment by Mr-Neutr0n** [comment]
I've opened a PR to fix this: #29153

The root cause is in `FCodePrinter._print_Function` where all function arguments are passed through `N(a, prec)`, which converts integer literals to floats. The fix skips the `N()` call for arguments that are integer-typed expressions (like `n + 1` where `n` is an integer symbol) while still allowing constant folding for pure numeric constants (like `log(10)`).
